cpp/splinepy/utils/arrays.hpp

```diff
@@ -0,0 +1,283 @@
+#pragma once
+
+#include <array>
+#include <cmath>
+#include <algorithm>
+#include <numeric>
+
+namespace splinepy::utils {
+
+using namespace splinelib::sources;  
+
+/// Elementwise subtraction
+template<typename T, typename NamedType, std::size_t dim>
+inline void FirstMinusSecondEqualsThird(
+    const std::array<NamedType, dim>& first,
+    const T* second, /* c array */
+    std::array<T, dim>& third) {
+  for (size_t i{}; i < dim; ++i) {
+    // following line should raise error during compile time
+    // if T != NameType::Type_
+    third[i] = static_cast<T>(first[i]) - second[i];
+  }
+}
+
+/// Elementwise mean
+template<typename T, std::size_t dim>
+inline std::array<T, dim> Mean(const std::array<T, dim>& arr1,
+                               const std::array<T, dim>& arr2) {
+  std::array<T, dim> out;
+  for (int i{0}; i < dim; ++i) {
+    out[i] = (arr1[i] + arr2[i]) * .5;
+  }
+  return out;
+}
+
+/// Elementwise mean overload, you can specified return value_type.
+/// Useful for returning NamedType.
+template<typename ReturnT, typename T, std::size_t dim>
```

Comment:
```suggestion
template<typename T, std::size_t dim, typename ReturnT=T>
```
And you can delete the other function

Reply:
err